src/main/scala/viper/gobra/translator/encodings/maps/MapEncoding.scala

```diff
@@ -0,0 +1,364 @@
+          vP <- goE(perm)
+        } yield vpr.FieldAccessPredicate(vpr.FieldAccess(vE, underlyingMapField(ctx))(pos, info, errT), vP)(pos, info, errT)
+    }
+  }
+
+  override def finalize(col: Collector): Unit = {
+    if (isUsed) {
+      col.addMember(genDomain())
+    }
+  }
+
+  private var isUsed: Boolean = false
+
+  private val keyParam = vpr.TypeVar("K")
+  private val valueParam = vpr.TypeVar("V")
+
+  /**
+    * Generates
+    *   domain GobraMap[K,V] {
+    *     function getMap(addr: Int): Map[K,V]
+    *   }
+    */
+  private def genDomain(): vpr.Domain = {
+    vpr.Domain(
+      name = domainName,
+      functions = Seq(getMapFunc),
+      axioms = Seq(),
+      typVars = Seq(keyParam, valueParam)
+    )()
+  }
+
+  private val getMapFuncName: String = "getMap"
+  private val getMapFunc: vpr.DomainFunc = vpr.DomainFunc(
+    name = getMapFuncName,
+    formalArgs = Seq(vpr.LocalVarDecl("id", vpr.Int)()),
+    typ = vpr.MapType(keyParam, valueParam),
+  )(domainName = domainName)
+
+  /**
+    * Field of the corresponding map id
+    */
+  private def underlyingMapField(ctx: Context): vpr.Field = ctx.field.field(in.IntT(Exclusive))(ctx)
+
+  /**
+    * Builds the expression `getMap([exp].underlyingMapField)`
+    */
+  private def getCorrespondingMap(exp: in.Expr, keys: in.Type, values: in.Type)(ctx: Context): CodeWriter[vpr.Exp] = {
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+    def goT(t: in.Type): vpr.Type = ctx.typeEncoding.typ(ctx)(t)
+
+    for {
+      vExp <- goE(exp)
+      res = withSrc(vpr.DomainFuncApp(
+        func = getMapFunc,
+        args = Seq(withSrc(vpr.FieldAccess(vExp, underlyingMapField(ctx)), exp)),
+        typVarMap = Map(keyParam -> goT(keys), valueParam -> goT(values))
+      ), exp)
+    } yield res
+  }
+
+  /**
+    * Builds the expression `idx in Domain(vprMap)`
+    */
+  private def goMapContains(vprMap: vpr.Exp, idx: vpr.Exp)(src: in.Node): vpr.Exp =
+    withSrc(vpr.AnySetContains(idx, withSrc(vpr.MapDomain(vprMap), src)), src)
+
+  /**
+    * Computes the result of looking up a value in an indexed expression and a bool expression asserting
+    * whether the key is in the map
+    */
+  private def goMapLookup(lookupExp: in.IndexedExp)(ctx: Context): CodeWriter[(vpr.Exp, vpr.Exp)] = {
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    lookupExp match {
+      case l@in.IndexedExp(exp :: ctx.Map(keys, values), idx) =>
+        for {
+          vIdx <- goE(idx)
+          isComp <- MapEncoding.checkKeyComparability(idx)(ctx)
+          vDflt <- goE(in.DfltVal(values)(l.info))
+          correspondingMap <- getCorrespondingMap(exp, keys, values)(ctx)
+          containsExp = goMapContains(correspondingMap, vIdx)(l)
+          lookupRes = withSrc(vpr.CondExp(containsExp, withSrc(vpr.MapLookup(correspondingMap, vIdx), l), vDflt), l)
+          lookupResCheckComp <- assert(isComp, lookupRes, comparabilityErrorT)(ctx)
```

Comment:
what does this assert do here? Why do we need to check comparability of the key? Can this not be checked statically?

Reply:
No, at least not fully statically. If the static type of the key is an interface type, this check must be done at runtime